src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/srm/algebras/BDD.cs

```diff
@@ -461,72 +458,78 @@ public bool Contains(int input)
         /// </summary>
         public bool IsEssentiallyBoolean(out BDD terminalActingAsTrue)
         {
+            // True and False are not MTBDD leaves.
             if (IsFull || IsEmpty)
             {
                 terminalActingAsTrue = null;
                 return false;
             }
-
-            if (IsLeaf)
+            else if (IsLeaf) // Otherwise any leaf is an MTBDD leaf.
             {
                 terminalActingAsTrue = this;
                 return true;
             }
 
-            var stack = new Stack<BDD>();
-            var set = new HashSet<BDD>();
+            // This will hold the unique MTBDD leaf.
+            BDD leaf = null;
 
-            stack.Push(this);
+            // Maintain sets of nodes to visit and already visited.
+            var toVisit = new Stack<BDD>();
+            var visited = new HashSet<BDD>();
 
-            BDD leaf = null;
+            toVisit.Push(this);
 
-            while (stack.Count > 0)
+            // Consider all nodes to find an MTBDD leaf and check that it is unique (apart from False).
+            while (toVisit.Count > 0)
             {
-                var node = stack.Pop();
+                var node = toVisit.Pop();
                 if (node.IsEmpty)
                     continue;
 
                 if (node.IsFull)
                 {
-                    //contains the True leaf
+                    // Contains the True leaf, so not essentially boolean in this method's sense.
```

Comment:
Another nit not worth fixing globally. We seem to usually start inline comments with lower case.